src/storage/sql/mod.rs

```diff
@@ -1 +1,319 @@
+        snippet_id: i32,
+        snippet: &Snippet,
+    ) -> Result<(), StorageError> {
+        diesel::delete(
+            tags::table.filter(
+                tags::snippet_id
+                    .eq(snippet_id)
+                    .and(tags::value.ne_all(&snippet.tags)),
+            ),
+        )
+        .execute(conn)?;
+
+        Ok(())
+    }
+}
+
+impl Storage for SqlStorage {
+    fn create(&self, snippet: &Snippet) -> Result<Snippet, StorageError> {
+        let conn = self.pool.get()?;
+        conn.transaction::<_, StorageError, _>(|| {
+            // insert the new snippet row first to get the generated primary key
+            let snippet_id = self.upsert_snippet(&conn, snippet)?;
+            // insert the associated changesets
+            self.upsert_changesets(&conn, snippet_id, snippet)?;
+            // insert the associated tags
+            self.upsert_tags(&conn, snippet_id, snippet)?;
+
+            Ok(())
+        })?;
+
+        // reconstruct the created snippet from the state persisted to the database
+        self.get(&snippet.id)
+    }
+
+    fn get(&self, id: &str) -> Result<Snippet, StorageError> {
+        let conn = self.pool.get()?;
+        conn.transaction::<_, StorageError, _>(|| {
+            let result = snippets::table
+                .filter(snippets::slug.eq(id))
+                .get_result::<models::SnippetRow>(&conn);
+            let snippet = match result {
+                Ok(snippet) => snippet,
+                Err(diesel::NotFound) => return Err(StorageError::NotFound { id: id.to_owned() }),
+                Err(e) => return Err(StorageError::from(e)),
+            };
+
+            let changesets = changesets::table
+                .filter(changesets::snippet_id.eq(snippet.id))
+                .get_results::<models::ChangesetRow>(&conn)?;
+            let tags = tags::table
+                .filter(tags::snippet_id.eq(snippet.id))
+                .get_results::<models::TagRow>(&conn)?;
+
+            Ok(Snippet::from((snippet, changesets, tags)))
+        })
+    }
+
+    fn update(&self, snippet: &Snippet) -> Result<Snippet, StorageError> {
+        // load the snippet from the db to check if we need to update anything
+        let persisted_state = self.get(&snippet.id)?;
+        if persisted_state == *snippet {
+            // if not, simply return the current state
+            Ok(persisted_state)
+        } else {
+            // otherwise, potentially update the title and the syntax
+            let conn = self.pool.get()?;
+            conn.transaction::<_, StorageError, _>(|| {
+                let snippet_id = self.upsert_snippet(&conn, snippet)?;
+                // insert new changesets and tags
+                self.upsert_changesets(&conn, snippet_id, snippet)?;
+                self.upsert_tags(&conn, snippet_id, snippet)?;
+                // and delete the removed tags
+                self.trim_removed_tags(&conn, snippet_id, snippet)?;
```

Comment:
**Q:** Maybe we should think of grouping these together?

Reply:
Not sure I understand what you mean by that. Grouping what?

Reply:
Just ignore that.

Reply:
Ok :)